haystack/core/serialization.py

```diff
@@ -59,7 +80,9 @@ def generate_qualified_class_name(cls: Type[object]) -> str:
     return f"{cls.__module__}.{cls.__name__}"
 
 
-def component_from_dict(cls: Type[object], data: Dict[str, Any]) -> Any:
+def component_from_dict(
+    cls: Type[object], data: Dict[str, Any], name: str, callbacks: Optional[DeserializationCallbacks] = None
```

Comment:
Why do we need the `name` here? That's not a concept the Component should be aware of if not strictly necessary, that's only related to the Pipeline ideally.

Reply:
The abstraction of the name doesn't get leaked into the `Component` class with this change? It's still contained within the serialization code. 

As to why it's needed, callbacks should be able to disambiguate components that occur multiple times in a pipeline but with different names.

Reply:
Ah, make sense. To be precise we're talking about the Component type not the instance here right? You can't add a Component multiple times, even when adding to the same Pipeline.

Reply:
Correct.